### DISCUSSION TITLE
Global Hotkeys On Main Thread

### DISCUSSION THREAD

**Discussion by CryptokidFH** [delfick/photons#75] [Q&A]
Hello, I'm very new to Python and was hoping someone could point me in the right direction.

I'm trying to write a basic script that will toggle certain lights when certain key events are pressed.
I'm using the Keyboard module to grab those events, and then using run(target:toggle etc.) to execute the commands. This approach works well enough for me, but there is a slight delay because the toggle operations need to perform discovery each time. I noticed I can reduce this delay if I use hardcoded discovery for my entire device list, but this means I will have to ensure my device ip list is always maintained.

How can I either:
1. Ensure the hardcoded list is maintained
2. Dynamically discover all lights once at run time, and then make the keyboard hotkeys run the toggle command without discovery

I've been attempting approach 2, and I've been reading through the docs but I'm definitely having trouble in some areas. When I try doing the library setup, my get_label function runs correctly, but if I try and use the keyboard hotkeys to call any other separate functions that make await operations, my console tells me the operation was never awaited and nothing happens. I'm not really sure what I'm doing wrong, I think its partially a general lack of understanding on how to use await properly?

Again I'm very sorry if I'm just misunderstanding the docs, very new to this.

**Comment by delfick** [comment]
> Again I'm very sorry if I'm just misunderstanding the docs, very new to this.

All good, we all start somewhere :)

> This approach works well enough for me, but there is a slight delay because the toggle operations need to perform discovery each time

Sounds like you're recreating the session each time.

Do you have the code on github or somewhere and I can take a look?

**Reply by CryptokidFH** [reply]
Very very rough, just added comments to make it more sensible:

https://haste.zneix.eu/osomowawil.py

**Reply by delfick** [reply]
yeap, so when you do

```
    async with lan_target.session() as sender:
        await accent_toggle(
            lan_target, accent
        )  # calling accent toggle like this works when I run the script, but calling it inside a hotkey doesn't work
```

You can pass in the sender instead of lan_target and then use the sender directly. The sender is a session object and will hold onto where everything is.

Also run_coro_as_main would indeed be better (in that it does what you're doing yourself):

```
if __name__ == "__main__":
    collector = library_setup()
    collector.run_coro_as_main(get_label(collector))
```

Also the keyboard module you're using isn't asyncio, which is why that doesn't work so well. I'd recommend python prompt toolkit instead https://python-prompt-toolkit.readthedocs.io/en/master/index.html

So something like this https://gist.github.com/delfick/58eccf482c71f187917a6e55b58bddaf

I got a bit carried away with that lol. Python prompt toolkit is great but it can be annoying sometimes. I didn't comment it very well though, so please don't be afraid to ask a million questions :)

**Reply by delfick** [reply]
and yes, when you run `li.run('lan:power_toggle match:group_name=Accent -- {"duration": 0.5}'))` you're starting an entire new instance of photons, which comes with the significant startup cost of photons as well as discovery.

**Reply by CryptokidFH** [reply]
So I've looked through and tested yours for a day and started working towards integrating into mine. I don't completely understand futures yet but I'm getting there, in our case its just an object representing the 'future' result of the main thread async operation for the collector right? This is why you wait for first future in the start function, since if it was returned it mains our main thread somehow finished? Which is not the quite the same as exiting our app with a shortcut, although both close the session similarly?

Your code works for me probably as you intended, but I have some work to do to make it work the way I want I think. The reason I used the keyboard module was because ideally I want this program to run in the background when Windows starts. I don't want any windows open or anything, I just want a background script that can grab my global hotkeys no matter what program is in focus (understandably, some full screen apps might mess with this) and toggle my lights. Do you know if Python prompt supports this kind of functionality, or should I look elsewhere? I tried looking through the documentation but I quickly found myself getting overwhelmed. I guess it would help if I understood how you were able to tell that Keyboard wasn't asyncio? 

This hotkey functionality has always been my primary goal, I took a long time playing with various languages and finally decided to pick up Python with Photons because it seemed like probably the best LAN implantation out there and like I said, I wanted responsiveness. That being said do you think Python is maybe just a bad choice for what I want to do? I'm happy to continue messing with this library and learning things in Python, I just want to make sure I'm not wasting my time going in the wrong direction with a language.

The other part I'm curious about, you mention having to refresh references with that function, and you have it set to call every 5 mins. Can I not call the ref.reset at the end of each hotkey event, so that the refs are updated automatically each time a light is changed? Eventually I plan on potentially integrating this with a music visualizer in some way (I've been dreaming of a lifx-milkdrop plugin for ages), so I want light states to be quickly updated probably right? Or am I misunderstanding the use of the ref resets?

Thank you very much for the help so far by the way, I definitely wasn't expecting you to write me a whole script haha.

**Reply by delfick** [reply]
> in our case its just an object representing the 'future' result

Yeap. The future as a concept is exactly that. It's a place where the value will go when it's ready. We can ask the future if it's done with `fut.done()` which returns True or False. A Future is done if it has a result, has an exception or is cancelled. If the future isn't done yet then we can try again later on, or we can tell the program we don't want to execute the next line in our code till it has that value by using ``await``.

In the async model that asyncio uses (co-operative concurrency), we can tell python that it can choose to do something else while we wait for a future value by using ``await``.

So practically, that means if I was to do

```
while True:
    if fut.done():
        break
    else:
        time.sleep(5)
```

Then it doesn't matter that we're using asyncio because this code won't ever let the python interpreter go execute something else.

Whereas

```
await fut
```

let's python choose to do something else while we wait for that future value.


> This is why you wait for first future in the start function, since if it was returned it mains our main thread somehow finished? Which is not the quite the same as exiting our app with a shortcut, although both close the session similarly?

In Photons I've made it so that `final_future` is only done when the program has ended. It let's me tell the code to wait till the program is finished before doing cleanup activities. In this case that particular line won't happen unless a `SIGTERM` signal has been sent to the app from outside. Given it's more likely a ctrl-c for example will be caught by the application itself.

```
await hp.wait_for_first_future(
    final_future, task, name="start::wait_for_exit"
)
```

Says "wait for either final_future or the task to finish, it doesn't matter which one is done first"

> The reason I used the keyboard module was because ideally I want this program to run in the background when Windows starts.

oh, hidden requirement!!

> I don't want any windows open or anything, I just want a background script that can grab my global hotkeys

Riiight, that explains why your shortcuts were so weird.

You essentially want a keylogger. Quick search and it seems the keyboard module is indeed what you want. The good news is I've thought of a simpler way of doing things with using keyboard so it's all good.

> I tried looking through the documentation but I quickly found myself getting overwhelmed.

Yeah, python prompt toolkit is a lot

> I guess it would help if I understood how you were able to tell that Keyboard wasn't asyncio?

One of the more annoying things about asyncio is to do anything asyncio requires the thing using it to be async and the thing using that to be async, etc. Which is why you can't await on photons code in a synchronous callback.

But what we can do is use keyboard to get the event and handle that event separately to those callbacks. In this way it's unlikely to matter that keyboard itself isn't using asyncio, because I think the nature of how it works itself is asynchronous. And as long as the callbacks we give it do very little then it's not overly different (because in python asyncio we can do only one thing at a time still)

> Python with Photons because it seemed like probably the best LAN implantation out there

:D

> That being said do you think Python is maybe just a bad choice for what I want to do?

Your bottleneck is the network. Photons has the ability to cache network stuff which is why if you press 'a' and then 'b' and then 'a' with my demo your lights should be changing to red and blue and red pretty instantly.

> The other part I'm curious about, you mention having to refresh references with that function, and you have it set to call every 5 mins. Can I not call the ref.reset at the end of each hotkey event, so that the refs are updated automatically each time a light is changed? Eventually I plan on potentially integrating this with a music visualizer in some way (I've been dreaming of a lifx-milkdrop plugin for ages), so I want light states to be quickly updated probably right? Or am I misunderstanding the use of the ref resets?

The way we do discovery in the LIFX ecosystem is we broadcast a special message to the network and wait for all the lights on the network to respond. Then we know what lights are on the network.

So for a reference that represents all the lights with the group name "Accent", means we've found all the lights and asked them what their group is and then hold onto that information. If one of the lights doesn't respond or you've turned it on since that discovery then we won't know about it! It's unlikely this is a common instance so every 5 minutes is probably often enough to check again. If you do it after each change then when you're making lots of changes you'll waste time waiting for discovery and if you're making rare changes then it's likely out of date again by the time you run a command again.

> Thank you very much for the help so far by the way, I definitely wasn't expecting you to write me a whole script haha.

You're welcome. I'm still working out how to use python-prompt-toolkit, so it was a fun little exercise :)

I mentioned above I have an idea. I don't have time right now to write full working code, but it'd look something like this (same disclaimer of don't be afraid to ask lots of questions applies):

```
async def toggle_accent_lights(sender):
    await sender(...)

async with hp.ResultStreamer(final_future) as streamer:
    queue = hp.Queue(final_future)

    async def get_events():
        async with queue:
            async for action in queue:
                # By adding it to the streamer, we run the action from this event in the background
                await streamer.add_coroutine(action(sender))
        
    await streamer.add_coroutine(get_events())

    # Calling this means that the streamer will finish when there's no more tasks left
    # Which will be when all the hotkeys have been processed and the queue thinks it's done
    # Which is when final_future is done.
    streamer.no_more_work()

    # The queue let's us synchronously add actions in the callback
    # And our get_events coroutine is able to start those actions in the streamer
    # which requires to be done in an async function because of how the streamer is implemented

    keyboard.add_hotkey('ctrl+alt+shift+up', lambda: queue.append(toggle_accent_lights))
    keyboard.add_hotkey(...)
    keyboard.add_hotkey(...)
    keyboard.add_hotkey(...)

    async for result in streamer:
        if not result.successful:
            print(result.context, result.value)
```

**Reply by delfick** [reply]
and if you wanted to append an action with arguments, then you can use partial like this

```
from functools import partial

async def toggle_lights(reference, sender):
   await sender(PowerToggle(), reference)

queue.append(partial(toggle_lights, accent_lights))
```

using partial like that is equivalent to doing something like

```
async def toggle_accent_lights(sender):
   return await toggle_lights(accent_lights, sender)

queue.append(toggle_accent_lights)
```

**Reply by CryptokidFH** [reply]
So I'm pretty sure I follow what you're doing here but I'm a bit confused on what I do in my start function with this new result streamer compared to the old task holder you wrote? Do I scrap the task holder completely or am I supposed to still be using it to initiate the tasks for the queue? I've been trying to set up the queue like you have but I keep getting an AttributeError __aexit__ with queue in get_events.

Here's where I'm at:
https://haste.zneix.eu/amyvureseb.py

**Comment by delfick** [comment]
Oh, it seems I didn't implement Queue as a context manager. This surprises me, I'll fix that another day.

So with apologise for the whiplash I'll say ignore the queue and result streamer. I can be a bit overly eager to use them sometimes cause they're very cool. My brain was a bit scattered yesterday and on reflection we don't need those. 

You should be able to get away with something like this https://gist.github.com/delfick/f6899cccb79ceed10eb7d1b0cda286ef

Though I can't run it because when I import keyboard it crashes ahhahah

Also your version was very close but it wasn't creating the sender right. It needs to be started and shutdown which is what the [`async with` syntax is doing.](https://book.pythontips.com/en/latest/context_managers.html)

**Reply by CryptokidFH** [reply]
So I definitely follow this one the most so far, especially considering it was somewhat close to what I was trying to do when I was originally modifying the task holder example you previously sent, and I'm starting to understand the module a bit better overall now.

The problem I'm having now is, every time I try and call a hotkey, I'm told there is no current event loop in the thread 'Thread-8'. Its thread 8 every time but I'm not sure that matters.

If I had to guess, this dev comment in Keyboard probably explains it:

Note: the callback is executed in a separate thread, asynchronously. For an
    example of how to use a callback synchronously, see `wait`.

I'm assuming this means that despite trying to make the hotkey in our main thread, every time we execute it, the callback function is trying to execute in a new thread instead of our main event loop, meaning it doesn't know anything about our session? Do you think I need to modify the way keyboard creates these callback tasks so their added the photons thread instead of their own? These seems like it would be a hassle so instead I'm currently looking into the different ways I can pass the callback to a different thread.

Maybe I should just create autohotkey hooks for my script instead lol..

Thank you very much again for your help so far!

**Reply by delfick** [reply]
> So I definitely follow this one the most so far

:)

> especially considering it was somewhat close to what I was trying to do when I was originally modifying the task holder example you previously sent, and I'm starting to understand the module a bit better overall now.

Yay. Yeah I got a bit carried away for a bit with the fancy things in Photons.

> If I had to guess, this dev comment in Keyboard probably explains it:

> Note: the callback is executed in a separate thread, asynchronously. For an
> example of how to use a callback synchronously, see wait.

Yeah. That makes sense.

Let's do something like this instead

```
ts.add(refresh_references(sender, ts))

loop = asyncio.get_event_loop()

def add_hotkey(key, action):
    keyboard.add_hotkey(key, lambda: loop.call_soon_threadsafe(action))

add_hotkey(
    "ctrl+alt+shift+up",
    lambda: ts.add(power_toggle(accent_lights)),
)

print("listening")
await final_future
```

I don't have the time to run that right now but I think it should work.

**Reply by CryptokidFH** [reply]
Success! Thank you very much for your help, everything seems to be working as intended and I definitely learned a few things, here's what I ended up with: https://haste.zneix.eu/eqiwusodyf.py

I modified the add hotkey to add toggle since every hotkey I was having to pass sender to the toggle function, figured it was easier to pass it in just the one function.

I tried to make an exit shortcut a few different ways but it seems to be the last thing I'm stuck on. Calling exit stops my shortcuts from running, but it leaves a python process in the background, I'm assuming because photons doesn't get to do proper cleanup. I also tried stop_photons_app a few different ways, but usually I either end up with coroutine not awaited or a CancelledError Exception on await final_future. Similar results with ts.finish. I assume I will never recieve a final future because there will always be a refresh_references task right?

**Reply by delfick** [reply]
> Success!

wooh!!

> Thank you very much for your help

You're very welcome :)

> I modified the add hotkey to add toggle since every hotkey I was having to pass sender to the toggle function

Oh cool, that's awesome

> I tried to make an exit shortcut a few different ways but it seems to be the last thing I'm stuck on.

Try

```
keyboard.add_hotkey("ctrl+alt+shift+left+right", lambda: loop.call_soon_threadsafe(final_future.cancel))
```

**Reply by CryptokidFH** [reply]
Same CancelledError Exception: no description
on await final_future
:S

I really do want to attempt a music visualizer next, but I feel these threads are gonna be the death of me first.

**Reply by delfick** [reply]
change that to

```
await hp.wait_for_first_future(final_future, name="start:wait_till_end")
```

> I really do want to attempt a music visualizer next

Good luck. Those are very difficult :)

**Reply by CryptokidFH** [reply]
Same exception, same line >_>

**Reply by delfick** [reply]
<img width="1008" alt="Screen Shot 2021-10-21 at 5 40 46 pm" src="https://user-images.githubusercontent.com/109301/138224873-fddce49f-d804-4f2c-a547-a838abe39ef4.png">

do you mean like that? That is normal because the application was cancelled.

In this case try this instead

```
                loop.call_soon_threadsafe(final_future.set_result, True)
```

**Reply by delfick** [reply]
unless you kept the `await final_future` you should replace that with the hp.wait_for_first_future.

The wait_for_first_future will swallow the exception, whereas await final_future will not.

**Reply by CryptokidFH** [reply]
Gahhhh I had been running it in debug mode and didn't realize the way it was displaying that wasn't a true error, how embarrassing :S

Everything seems to be working fine now. Thank you very much again for bearing with me while I learn python at the same time as your wonderful lib, I've learned a lot from this. I suppose to that now you can have your own hotkeys for your lights if you want them haha.

My final question would probably be about memory consumption. I'm currently sitting at 51MB when I first execute the script, do you think there's any way to trim this down at all? I've also noticed that the more commands I issue the more memory it uses, just steadily over time, but spamming a toggle can eventually ramp it up to 70MB or higher. I have plenty of ram and don't mind all that much but it's definitely a tad bit high for a light control shortcut lol.

**Reply by delfick** [reply]
> Gahhhh I had been running it in debug mode and didn't realize the way it was displaying that wasn't a true error

hehehe
> Everything seems to be working fine now. Thank you very much again for bearing with me while I learn python at the same time as your wonderful lib, I've learned a lot from this

super happy to help

> My final question would probably be about memory consumption. I'm currently sitting at 51MB when I first execute the script, do you think there's any way to trim this down at all? I've also noticed that the more commands I issue the more memory it uses, just steadily over time, but spamming a toggle can eventually ramp it up to 70MB or higher. I have plenty of ram and don't mind all that much but it's definitely a tad bit high for a light control shortcut lol.

A big chunk of Photon's memory use is related to python. On today's computers even 100MB is tiny and it's not really something that needs to be worried about.

When it climbs does it ever stop climbing and go down eventually?

**Reply by CryptokidFH** [reply]
> When it climbs does it ever stop climbing and go down eventually?

Yeah, when I put windows to sleep and wake again it was down to 3.9, went up to 9 after running some hotkeys, back to 50 after relaunching. Maybe some unnecessary stuff gets cleaned up when being put to sleep?

I'm definitely really happy to finally get this initial LAN project done, its been on my mind forever, and now that I've got an initial springboard I can finally start working on the real project I've wanted to do for years, the visualizer.

> Good luck. Those are very difficult :)

I'm pretty confident that I shouldn't have too hard a time with the audio processing for visuals, the biggest challenge I want to solve that I haven't seen other Lifx visualizers do, is properly calibrate the audio buffer for the average network latency. Every other Lifx visualizer I ever saw either suffered from significant delay or looked really boring / basic. I think the delay calibration could be done by listening to a virtual audio device by adding a buffer to a cable and then having python listen to the original source, but this solution feels messy and I don't like users needing external tools to run it. I guess I have a lot of reading to do on wasapi / windows audio lol. I might have to write a custom audio processing object to add the delay to the driver so python has time to read the buffer (assuming I can even get python to read the input buffer before the sound plays.

My final question which probably needs its own discussion ties to what I said near the end there about end user simplicity. Have you ever tried compiling anything using photons into a single .exe? I gave it a quick test with pyinstaller but it threw an error stack on: 
`executor > ... > messages > inspector in getsource / getsourcelines / findsource`, with 
` File "<frozen importlib._bootstrap>", line 991, in _find_and_load`
  `File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked`
 ` File "<frozen importlib._bootstrap>", line 671, in _load_unlocked`
between each module. I've not yet to tried other compile utils since it was just a quick test. I only ask because ideally, if I make that visualizer, to keep distribution user friendly I think it would be nice if they could just run a simple .exe and maybe edit a config file. Before I learned any python I often found executing a project I'd found to be less intuitive than projects in other langs, mainly due to confusion about proper env setup at the time, so I'd like to avoid that for newer users. Again not a huge priority, was just wondering if you knew anything about it.

**Comment by delfick** [comment]
> I'm definitely really happy to finally get this initial LAN project done, its been on my mind forever, and now that I've got an initial springboard I can finally start working on the real project I've wanted to do for years, the visualizer.

yay!

> I'm pretty confident that I shouldn't have too hard a time with the audio processing for visuals, the biggest challenge I want to solve that I haven't seen other Lifx visualizers do, is properly calibrate the audio buffer for the average network latency

This sounds very cool :)

> assuming I can even get python to read the input buffer before the sound plays.

I've never done anything with audio processing before. If there isn't already some C extension for this kind of stuff, you might end up needing to look at that making one. Super cool if turns out Python would be fast enough though :)

> Have you ever tried compiling anything using photons into a single .exe?

I've never really tried, but mainly cause it's not something I think would work.

What I've come to is how I explain to people how to run photons

```
python3 -m venv photons
./photons/bin/python -m pip install lifx-photons-core
./photons/bin/lifx lan:transform -- '{"power": "off"}'
```

So create an entry point console script and suggest people create virtualenvs to install photons into.

**Reply by delfick** [reply]
Though that approach can get tricky with C extensions and requires some familiarity with the console.
